search functionality to discuss

InteriorNature opened the pull request:
There are two approaches - it would be nice to integrate the functionality. 
1) <Geocoder> is a wrapper that shows the first searchbox. Uncomment DeckGL to see how it works.
 To move to a country by location, you can type in the country name into the upper left searchbox. DeckGL makes our searchbox disappear.
2) Comment out DeckGL, and you should see our searchbox which is a manual dropdown system rather than downshift or autosuggest or other 3rd party design. Type the letter N and a highlighted dropdown will show one item. Type A, and it will show 3 items. I didn't want to finish debugging with the temp database design, until we have a database design.

InteriorNature committed:
search functionality to discuss

InteriorNature committed:
search functionality/deck.gl added to demo

InteriorNature committed:
package.json update

InteriorNature committed:
deck.gl import removed

InteriorNature committed:
geojsonlayer

ArnasDickus commented on src/pages/Home/Map.js:
```diff
@@ -244,6 +286,24 @@ class Map extends Component {
+                        mapboxApiAccessToken={TOKEN}
+                        position="top-left"
+                    />
+                    {/* <DeckGL {...this.state.viewport} layers={[this.state.searchResultLayer]} /> */}
```
I tried using Deckgl input this is my steps:
Uncommenting Deckgl would give me an error DeckGL not defined. You also need to add this line for it to work.
`import DeckGL, { GeoJsonLayer } from 'deck.gl';`.  

Anyways, as you said one box disappeared, However, I see a few problems with input:
1) Clicking on input after one second it would deselect input. And in order to write, I need to click on input again.
2) Trying to type into input a few letters, it would reset form immediately. And I would need to select it again.

I did manage to make it work a few times but not enough to test it properly. You need to fix this bug first.

InteriorNature replied:
I only left that in for discussion. I noticed a significant delay before its suggestion list started to pop up. That is a wrapper created by a third party. The question is - should the backend include mapbox's way to search by location? If so, how do we do a location search versus a friends search by country? Which is more important?

InteriorNature replied:
Ran npm run format.

ArnasDickus commented on src/components/Search.js:
```diff
@@ -0,0 +1,214 @@
+import React, { Component, Fragment } from "react";
+import PropTypes from "prop-types";
+//import debounce from 'lodash.debounce';
```
If you not using debounce, I suggest removing a comment, and npm package.

InteriorNature replied:
Removed debounce comment.

ArnasDickus commented on src/components/Search.js:
```diff
@@ -0,0 +1,214 @@
+                    friend.country.toLowerCase().indexOf(userInput.toLowerCase()) > -1
+            ));
+            break;
+        // How are we storing interests? We will need a parser here.
```
Currently, we are not sure how we store interests. In [PR #84](https://github.com/zero-to-mastery/mappypals/pull/84) I raised concern. My suggestion was either:
1) We create or find a huge JSON file with all possible interest. So user types a few letters and a dropdown appears with all the possible interest, that he can select it. Playing video games, playing basketball, etc. 
2) We expand the interest section. So if a user clicks on playing video games another input appears saying what video games. using maybe IMDB API database we could allow the user to choose the specific game.

ArnasDickus commented on src/pages/Home/Map.js:
```diff
@@ -3,6 +3,11 @@ import MapGL, { Marker } from 'react-map-gl';
+import { GeoJsonLayer } from 'deck.gl';
+import Geocoder from 'react-map-gl-geocoder';
+import './mapbox-gl-geocoder.css';
+import FriendSearch from '../../components/Search'
```
This is kinda confusing. I suggest using the same class name to file. So FriendSearch class should be in FriendSearch.js file.

InteriorNature replied:
Changed to Friendsearch.

ArnasDickus commented on src/pages/Home/Map.js:
```diff
@@ -155,6 +164,38 @@ class Map extends Component {
+    };
+    
+    handleOnResult = event => {
+        console.log(event.result);
```
Unless this provides crucial information to the developer. I suggest removing console.log

InteriorNature replied:
Removed console.log.

ArnasDickus commented on src/pages/Home/mapbox-gl-geocoder.css:
```diff
@@ -0,0 +1,242 @@
+  background-color: transparent;
+  margin: 0;
+  height: 50px;
+  color: #404040; /* fallback */
```
Do we need to add fall back color if create-react-app already uses prefixes? [what is prefixes?](https://www.lifewire.com/css-vendor-prefixes-3466867)

InteriorNature replied:
Good question - that CSS is from the third party. I know prefixes are still needed for transition properties, but I would have to check into the others.

InteriorNature replied:
The fall back color in this case is used to make sure text remains visible while backgrounds are loading. It is not a prefix issue.

ArnasDickus reviewed (commented):
Hello, this is second biggest pull request submitted after [major release](https://github.com/zero-to-mastery/mappypals/pull/23). Anyways I have a few concerns here.

1) Your code is not prettified. Either download and use the pretify plugin. Or run a command `npm run format`

So from my understanding: 
Search.js - This is a search bar in the center. The user selects either country or interest and inputs text. That text would be filtered and compared to filteredFriends aka fake database.
1) I encountered a wierd bug, sometimes If I type anything I keep getting No Friends... message. And I can type anything and still get the same message. Sometimes I type letter n get friends like you suggested, but if I type q, the application crash. `Cannot read property 'firstname' of undefined let first = String(friend[index].firstname);`

SearchBar.js - styling css component
map.js - added Search.js and Deckgl input. So are you planning on using only one?

InteriorNature committed:
pull request comments fixed

ArnasDickus commented:
@InteriorNature So I see you fixed all problems.

Regarding your question. I don't know the answer. I need to ask @mubarakshow about this.
The question is - should the backend include mapbox's way to search by location? If so, how do we do a location search versus a friend search by country? Which is more important?

So we have 2 choices:
1) I prefer to merge this pull request because is already pretty big and create a new task when mubarakshow or marco respond to your question.
2) Keep this PR open until admins respond and when you can try completing the task fully 
What do you think?

InteriorNature committed:
searchitem crashes fixed

InteriorNature commented:
@ArnasDickus @mubarakshow I fixed the search bugs you were having in the friends search part. The only thing I couldn't fix was having text-overflow: ellipsis work for extra long names -I added a long name to the test database (type "n" to see what happens). I think it may be ready to merge. In terms of the geocoding search box - even if we do not have the backend database design figured out - a backend task could be added to figure out how to geocode and update the result of the friends search by country. The backend coder can remove the existing geocoder, since the wrapper isn't the best, and maybe come up with something better.
